compiler/rustc_mir/src/interpret/place.rs

```diff
@@ -235,9 +226,11 @@ impl<'tcx, Tag: Copy> MPlaceTy<'tcx, Tag> {
     }
 
     #[inline]
-    pub(super) fn vtable(&self) -> Scalar<Tag> {
+    pub(super) fn vtable(&self) -> AllocId {
         match self.layout.ty.kind() {
-            ty::Dynamic(..) => self.mplace.meta.unwrap_meta(),
+            ty::Dynamic(..) => {
+                self.mplace.meta.unwrap_meta().assert_ptr_alloc_id_with_zero_offset()
```

Comment:
I don't think this is right. libstd could dynamically create vtables via a heap allocations or promotion. These vtables would not have to have zero offsets. There are also ideas for vtable generation where multiple vtables are in one allocation. This here hardcodes something that is fairly easy to trigger from user code.

So we should keep the whole Scalar<Tag> thing here, or at best move to Pointer<Tag> (but that should be a separate PR imo).